<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

Признаки: пол, возраст и зарплата застрахованного, количество членов его семьи.

Целевой признак: количество страховых выплат клиенту за последние 5 лет.

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
data=pd.read_csv('/datasets/insurance.csv')
data.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

Изучим наш дата фрейм

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [4]:
data.isnull().sum()

Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64

Пропущенных данных нет

In [5]:
data.duplicated().sum()

153

В таблице наблюдаются явные дубликаты. Они составляют 

In [6]:
153/5000*100

3.06

3.06% от всех данных, так что можем удалить их

In [7]:
data.drop_duplicates(inplace=True)

In [8]:
data.duplicated().sum()

0

Избавились от дубликатов

**Вывод**

В первой части нашей работы выполнили предобработку данных, обнаружили дубликаты и удалили их. Теперь можем перейти к следующему шагу.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** качество линейной регрессии не изменится

**Обоснование:** 
Рассмотри случай с a'. По условию признаки умножаем на обратимую матрицу 
$$
a'=X'w', где X'=XP
$$
$$
w' = (X'^T X')^{-1} X'^T y
$$
подставим X' в w'
$$
w' = ((XP)^T (XP))^{-1} (XP)^T y
$$
теперь w' в а'
$$
a'=(XP)((XP)^T (XP))^{-1} (XP)^T y)
$$
$$
a'=(XP)(P^T X^T XP)^{-1} (XP)^T y)
$$
$$
a'=(XP)((P^T)^{-1} (X^T X)^{-1} (P)^{-1} P^T X^T y)
$$

Воспользумеся свойством 

$$
PP^{-1}=E
$$

аналогично для транспорированных матриц 
$$
P^T(P^T)^{-1}=E
$$

$$
a'=(XP)((X^T)^{-1} X^{-1} (P)^{-1} X^T y)
$$
$$
w = (X^T X)^{-1} X^T y
$$
$$
a'=(XP)(w(P)^{-1})
$$
$$
a'=XwP(P)^{-1}=Xw=a
$$

**Что и требовалось доказать**

**Вывод**

Доказали, что при умножении признаков на обратимую матрицу качество не изменится

1) С помощью математических преобразований найдём соотношение между $w$ и $w_p$. В моих условиях $w$ и $w'$
$$
w' = ((XP)^T (XP))^{-1} (XP)^T y
$$
$$
w' = (P^T X^T XP)^{-1} P^T X^T y
$$
$$
w' = P^{-1}(X^T X)^{-1} (P^T)^{-1} P^T X^T y
$$
$$
w' = P^{-1}(X^T X)^{-1} X^T y
$$
$$
w = (X^T X)^{-1} X^T y
$$
$$
w'=P^{-1}w
$$

2) Воспользуемся подсказкой расскроем выражение $((XP)^T XP)^{-1}$

$$
((XP)^T XP)^{-1}=(P^T X^T XP)^{-1}= P^{-1}(X^T X)^{-1} (P^T)^{-1}
$$

Из полученного выражения видно,нужно чтобы матрицы $ P,(X^T X), (P^T)$ были квадратными. 
Рассмотрим, $(X^T X)$ независио от того какая матрица $X$ этим умножением мы получим квадратную матрицу, так, например, если матрица $X$ имеет размерность(3,2), то траспорированная будет (2,3), а их умножение (2,2). Отсюда следует, что нужно задать условие на матрицу $P$, чтобы она была квадратной. А матрица $X$ может иметь любую размерность 

## Алгоритм преобразования

**Алгоритм**
Будем использовать алгоритм, указанный выше: умножаем матрицу признаков на обратимую матрицу.


**Обоснование**

В пункте 2 доказали, что качество не изменяется при использовании данного алгоритма. Но так как мы умножаем случайную матрицу, то должны проверить на обратимость

## Проверка алгоритма

Сравним качество наших моделей с использованием преобразований данных и без

Выделим обучающую и тестовую выборки

In [9]:
data_train, data_test = train_test_split(data, random_state=12345)

Признаки: пол, возраст и зарплата застрахованного, количество членов его семьи.

Целевой признак: количество страховых выплат клиенту за последние 5 лет.

In [10]:
features_train = data_train.drop('Страховые выплаты', axis=1)
target_train = data_train['Страховые выплаты']
features_test = data_test.drop('Страховые выплаты', axis=1)
target_test = data_test['Страховые выплаты']

In [11]:
class LinearRegression:
    def fit(self, train_features, train_target):
        X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
        y = train_target
        w = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)
        self.w = w[1:]
        self.w0 = w[0]

    def predict(self, test_features):
        return test_features.dot(self.w) + self.w0
    

Cначала проверим значения без преобразований

In [12]:
model = LinearRegression()
model.fit(features_train, target_train)
predictions = model.predict(features_train)
print('на обучающей r2 = ', r2_score(target_train, predictions))
predictions = model.predict(features_test)
print('на тестовой r2 = ', r2_score(target_test, predictions))

на обучающей r2 =  0.43215820285832096
на тестовой r2 =  0.4230772749214825


Теперь воспользуемся нашим алгоритмом защиты персональной информации

Мы должны проверить, что наша рандомная матрица обратима, те ее определитель не равен нулю

In [13]:
n=4
P = np.array([])
for i in range(1000):
    P = np.random.random((n, n))
    det = np.linalg.det(P)
    if det!=0:
        break

In [14]:


features_train = features_train @ P
features_test = features_test @ P
model = LinearRegression()
model.fit(features_train, target_train)
predictions = model.predict(features_train)
print('на обучающей r2 = ', r2_score(target_train, predictions))
predictions = model.predict(features_test)
print('на тестовой r2 = ', r2_score(target_test, predictions))

на обучающей r2 =  0.4321582026358557
на тестовой r2 =  0.4230777281921366


Результаты практически идентичны, значит наш алгоритм действительно работает

# Вывод

1) Провели предообработку данных: удалили дубликаты

2) Доказали что при умножении матрицы признаков на обратимую матрицу, качество линейной регресии не изменяется
3) Проверили алгоритм на практике: результат успешный.
    
    Без преобразований матриц значение на обучающей r2 =  0.43215820285832096, на тестовой r2 =  0.4230772749214825
    
    С преобразованиями - на обучающей r2 =  0.4321360880675721, на тестовой r2 =  0.42286699882435363

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования